### PR TITLE
Fix VAE example

### PR DESCRIPTION
**Pull request by RuiShu**
A number of changes:
1. Switch from Lambda to merge, otherwise code will not run in keras==1.0.5.
2. Rename z_log_std to z_log_var in order for the objective function to make sense
3. Adjust reparameterization trick to reflect use of z_log_var, not z_log_std
4. Remove epsilon_std, since (standard) VAE uses isotropic gaussian prior.
5. Re-balance the weighting of KL and reconstruction terms
6. Use adam instead of rmsprop
7. Increase hidden unit size to improve model
8. Increase batch size to speed up training

### CODE REVIEW

**Comment by bytec0d3** [comment]
Hello,
I tried the VAE example with Keras 1.0.5 but an Exception raised at line 36:
`z = Lambda(sampling)([z_mean, z_log_std])`

The exception is the following:
`Exception: The `get_output_shape_for` method of layer "lambda_1"" should return one shape tuple per output tensor of the layer. Found: [(16, 2), (16, 2)]`

SO: Ubuntu 16.04
Python: 3.5
Keras: 1.0.5 (installed via pip)

**Comment by RuiShu** [comment]
This problem can be fixed in keras 1.0.5 by switching from using 
```
z = Lambda(sampling, output_shape=(latent_dim,))([z_mean, z_log_std])
```
 to merge
```
z = merge([z_mean, z_log_var], mode=sampling, output_shape=(latent_dim,))
```
Digging into it a bit more, it seems that this is an issue with Lambda that has since been fixed in https://github.com/fchollet/keras/commit/229f13a8648596b1a2cbf96c92e087f821a8d6df

File: examples/variational_autoencoder.py [hunk 1]
```diff
@@ -40,8 +37,8 @@ def sampling(args):
 x_decoded_mean = decoder_mean(h_decoded)
 
 def vae_loss(x, x_decoded_mean):
-    xent_loss = objectives.binary_crossentropy(x, x_decoded_mean)
-    kl_loss = - 0.5 * K.mean(1 + z_log_std - K.square(z_mean) - K.exp(z_log_std), axis=-1)
+    xent_loss = original_dim * objectives.binary_crossentropy(x, x_decoded_mean)
```
Comment (lukovkin) [inline examples/variational_autoencoder.py]: @RuiShu Why do you think it's required to multiply by the original_dim?

Comment (RuiShu) [inline examples/variational_autoencoder.py]: The reconstruction term in the variational objective function is `E[- ln p(x|z)]`. Since we assume element-wise independence, this becomes `E[\sum -\ln p(x_i | z)]`. Thus, we need to sum the cross-entropy terms over all 784 dimensions. The implementation of `objectives.binary_crossentropy` averages over all dimensions (except the batch dimension). Thus, I multiplied.

File: examples/variational_autoencoder.py [hunk 2]
```diff
@@ -5,33 +5,30 @@
...
-                              mean=0., std=epsilon_std)
-    return z_mean + K.exp(z_log_std) * epsilon
+    z_mean, z_log_var = args
+    epsilon = K.random_normal(shape=(batch_size, latent_dim), mean=0.)
+    return z_mean + K.exp(z_log_var/2) * epsilon
 
 # note that "output_shape" isn't necessary with the TensorFlow backend
-# so you could write `Lambda(sampling)([z_mean, z_log_std])`
-z = Lambda(sampling, output_shape=(latent_dim,))([z_mean, z_log_std])
+z = merge([z_mean, z_log_var], mode=sampling, output_shape=(latent_dim,))
```
Comment (fchollet) [inline examples/variational_autoencoder.py]: Previous version was more readable imo.

File: examples/variational_autoencoder.py [hunk 3]
```diff
@@ -5,33 +5,30 @@
...
+z_log_var = Dense(latent_dim)(h)
 
 def sampling(args):
-    z_mean, z_log_std = args
-    epsilon = K.random_normal(shape=(batch_size, latent_dim),
-                              mean=0., std=epsilon_std)
-    return z_mean + K.exp(z_log_std) * epsilon
+    z_mean, z_log_var = args
+    epsilon = K.random_normal(shape=(batch_size, latent_dim), mean=0.)
+    return z_mean + K.exp(z_log_var/2) * epsilon
```
Comment (fchollet) [inline examples/variational_autoencoder.py]: PEP8

File: examples/variational_autoencoder.py (hunk 1 above)
Comment (lukovkin) [inline examples/variational_autoencoder.py]: I'm a bit confused here.
If T is the training set, T_i is the single i-th sample (image) from the training set (with a dim of 784) and t_j is the j-th pixel then when we consider p(x_i | z_i,l) (expr. 10), do we assume that x_i is a t_j (pixel) or T_i (image)?

But I could well be wrong, as I'm no expert in this field, I've started to look into VAE just recently.

Comment (RuiShu) [inline examples/variational_autoencoder.py]: `x^{(i)}` is a single training example (e.g. image) in Eq. 10 of Kingma's VAE paper. 

In Eq. 11, `x_i` refers to the i-th dimension of the training example (e.g. pixel).

Comment (lukovkin) [inline examples/variational_autoencoder.py]: @RuiShu Now I see the point, thank you. And the multiplication makes sense then, cool.

File: examples/variational_autoencoder.py (hunk 2 above)
Comment (RuiShu) [inline examples/variational_autoencoder.py]: I like the previous version too! Unfortunately, it doesn't run in keras 1.0.5 because this version contains the line `self.supports_masking = True` for Lambda. Should we just leave it be for when keras 1.0.6 rolls out?

File: examples/variational_autoencoder.py (hunk 3 above)
Comment (RuiShu) [inline examples/variational_autoencoder.py]: Put spacing around division sign? Just double checking :p

Comment (fchollet) [inline examples/variational_autoencoder.py]: Yes.

File: examples/variational_autoencoder.py [hunk 4]
```diff
@@ -40,8 +37,8 @@ def sampling(args):
 x_decoded_mean = decoder_mean(h_decoded)
 
 def vae_loss(x, x_decoded_mean):
-    xent_loss = objectives.binary_crossentropy(x, x_decoded_mean)
-    kl_loss = - 0.5 * K.mean(1 + z_log_std - K.square(z_mean) - K.exp(z_log_std), axis=-1)
+    xent_loss = original_dim * objectives.binary_crossentropy(x, x_decoded_mean)
+    kl_loss = - 0.5 * K.sum(1 + z_log_var - K.square(z_mean) - K.exp(z_log_var), axis=-1)
```
Comment (glenwayguy) [inline examples/variational_autoencoder.py]: Does the change to using sum instead of mean need to be made in variational_autoencoder_deconv.py as well, or is the difference between the functions intentional?

Comment (RuiShu) [inline examples/variational_autoencoder.py]: When using any fully factorized distribution `f(a|b)`, `log f(a|b)` should be `sum log(a_i | b)`. So in theory, we should be using sums for `variational_autoencoder_deconv.py` as well.

Comment (glenwayguy) [inline examples/variational_autoencoder.py]: Does this actually have a performance impact, or does in not matter in practice?

Comment (RuiShu) [inline examples/variational_autoencoder.py]: Depends on what you're measuring :)

If measuring the true ELBO, messing with the objective function is probably not a good idea.

If measuring the meaningfulness of representation and visual quality of image, you may perhaps find, empirically, that averaging is better than summing. But there are probably more principled ways of improving such performances.
